### ISSUE TITLE
Sendfile consideration

### ISSUE THREAD

**Issue by ducthangho** [h2o/h2o#671]
Hello.
I am considering to use H2O or nginx as a reverse proxy for a map tile server. Basically, each tile is a png image of 256x256, which represent a small part of the map. There might be millions of those images, storing in a hierarchy of folders in the form of   "z/x/y.png" where z is the zoom level, x,y are the index of the tile.

Usually when a user requests a tile, the web server will look to see if the corresponding png file exists, if it does, simply return it, otherwise, compute the image on-the-fly using mapnik. I would expect 90-95% would hit the cache (aka no computation needed). According to my understandings, nginx utilizes sendfile(), which is a zero-copy syscall, and H2o uses writev() to send static files like this. 

1. Do you think H2o has any advantages in my use cases. 

2. Why didn't you implement sendfile() in the first place

3. What do you think of Facebook Proxygen vs H2o in terms of performance.

Many thanks and looking forward to hearing from you

**Comment by kazuho** [comment]
> Do you think H2o has any advantages in my use cases. 

Yes.  Serving many image files concurrently is one of the areas in which you can gain a lot by moving from HTTP/1 to HTTP/2, and per my knowledge H2O is by far the most sophisticated and optimized open-source server implementation of the protocol.
A good demonstration can be found on https://www.symfony.fi/entry/compare-resource-loading-between-http-2-and-http-1-1.

> Why didn't you implement sendfile() in the first place

You cannot use sendfile unless you are serving files over HTTP/1 (plaintext).
In case of HTTP/2 or if TLS is involved, the server needs to read the file onto the memory to do encryption and/or to send them in HTTP/2 frames.

And in case of HTTP/1 without encryption, the merit of using sendfile has declined over the decade.  Modern CPUs have a big internal cache memory; contents of temporary buffer no longer falls down to main memory and therefore the cost of one-copy has become marginal.

> What do you think of Facebook Proxygen vs H2o in terms of performance.

IIRC Proxygen does not support HTTP/2.

**Comment by ducthangho** [comment]
Many thanks Kazuho.
I have seen your example of serving multiple images using http2/http1.1.  However, I think the main advantage of http2 in this case was due to the fact that http2 utilizes the network better, i.e. using fewer connections to the same site. 
For a png file, I don't think further encryption/compression would bring any benefits. Isn't it better to just send the image (without even loading to userspace memory, no CPU consumption is needed) than loading+encrypting/compressing+sending.  Even in the case of http2, I still think it is best to just compress the header, then send the file as is, no further processing is needed. What do you think about this?

I am thinking of using libh2o to write a custom handler to generate png files on-the-fly, in case the images are not already cached. However, examples of libh2o are very few. Can you give me a hint at where to start?

On another note, I used to work on a project to build a in-memory cache database, similar to memcache, redis, aerospike,..., but faster. I came across a project called Seastar (https://github.com/scylladb/seastar) in which I used as my underlying framework. The key idea here is to use an Intel technology called DPDK to bypass kernel altogether and let the application to handle packets. By doing that the server can serve up to 10 millions concurrent connections at the same time while using Linux kernel (aka epoll), the upper limit would be 10k connections (see the C10k and C10M problems). I think if you want to make H2o an ultimate beast in terms of performance, you might need to bring stuff like bypassing kernel to the game. Seastar also have its own Http server although it is far from production ready. The initial benchmark vs nginx were very promising (better than nginx for  toy problems even without DPDK, serving a single html file). What do you think of the potential of building HTTP server on top of a bypassing kernel layer.

**Comment by kazuho** [comment]
> I have seen your example of serving multiple images using http2/http1.1. However, I think the main advantage of http2 in this case was due to the fact that http2 utilizes the network better, i.e. using fewer connections to the same site. 

To be precise, the difference comes from the fact that with HTTP/2 allows much more requests to be sent concurrently when compared to HTTP/1.1 (100 vs. 6).

> Even in the case of http2, I still think it is best to just compress the header, then send the file as is, no further processing is needed.

Right.  However the server still needs to load the contents of the file to divide them into HTTP/2 frames and to encrypt them (use of TLS is a de-facto requirement for HTTP/2).

> I am thinking of using libh2o to write a custom handler to generate png files on-the-fly, in case the images are not already cached. However, examples of libh2o are very few. Can you give me a hint at where to start?

Considering the fact that the estimated cache hit ratio is high and under the assumption that image generation is computationally intensive, I'd suggest using a pair of file handler and fastcgi handler.

For example, if you write a configuration like the following, H2O will first search for a file in the directory specified using `file.dir` directive, and serve the file if found.  Otherwise, the request will be delegated to the FastCGI hander.  Your image generator can run as a FastCGI application and return the generated image at the same time storing it under `/path/to/images`, so that following requests can read the file.
```
file.dir: /path/to/images
fastcgi.connect:
  port: /path/to/fcgi.sock
  type: unix
```
You can also implement a caching proxy using the mruby handler; the skelton code can be found at https://gist.github.com/kazuho/e1e91c82876ce71cece1.

> The key idea here is to use an Intel technology called DPDK to bypass kernel altogether and let the application to handle packets.

Thank you for the suggestion.

I am definitely looking into if we can use DPDK (or user-space TCP stack).

To be honest, the number of packet transmissions has become less in HTTP/2 when compared to HTTP/1 (since multiple responses can be sent in a single TCP packet) so there will be less benefit in using user-space TCP in terms of server-side performance.  However, to provide best responsiveness, we would like to obtain various parameters from the TCP stack (i.e. exact size of the available send window) at the moment we try to write to a socket; such information cannot be obtained (or tweaked) unless we use user-space TCP.

**Comment by mingodad** [comment]
This thread bring an interesting problem that h2o have somehow ignored so far "requests coalesce" and I hope h2o can implement it so it'll not kill/torture proxies with multiple identical requests, this adds need for shared data/state and if not done properly the server can starve doing nothing other than wait for shared data/state.

When kazuo suggest to implement the "FastCGI hander/application" and the h2o configuration he omits the problem of multiple identical requests not been coalesced, leaving for the applications to deal with it, if h2o implement and manage "requests coalesce" it'll make the life easier/safe for any application/handler, and in my point of view h2o is the one that should provide it instead of each application reinventing the wheel.

Cheers !

**Comment by ducthangho** [comment]
@mingodad: Spot on. I will eventually have to deal with multiple identical requests myself. 
@Kazuho: Many thanks, I will try your suggestions regarding handlers and FastCGI. 

According to my knowledge, DPDK is more than just a user-space TCP stack.
1. DPDK dedicated one/a few CPU cores to run just one task only, i.e. handling packets. No context switch, no cache-miss, no mutex, no interrupt, no blocking, no task scheduling, etc. The kernel cannot even interfere nor preempt the execution of the dedicated network cores. Packets coming to the NIC, will be forwarded immediately to the network cores (in a zero copy manner)  and  be processed as fast as possible (In fact, you cannot do better than that). Packets that require expensive computation are sent to another core using concurrent queue. Again no mutex is needed.

Modern computers are capable of processing packets at insane speed, the bottle-neck is always network IO. The reason is because Linux kernel is designed to manage shared resource, and therefore, using mutex internally. Linux kernel used mutex when copy data from/to its internal network buffer and mutex is not handling contention very well -> 10k concurrent connections is about the limit for Linux kernel (and for any TCP-based server). In high-load real-world server, each client connecting to the server requires a new connection. No matter how fast your HTTP server is, you can never serve more than 10k clients simultaneously.
2. An optimized user-space TCP stack can be faster than the default one (just like your HTTP parser is faster than the Apache/nginx parser)
3. By fixing cores for a single task, it is easier to implement a better memory allocation (NUMA-award per thread allocation --> again avoid allocation contention)

The down side of DPDK is that it requires rather specialized NIC. Fortunately, DPDK NICs are fairly common nowadays. 

All in all, I really think the combination of H2o and a low-level server-side framework like Seastar would be advantageous, and potentially make H2o the fastest HTTP server ever. Just my two cents.

**Comment by utrenkner** [comment]
> However the server still needs to load the contents of the file to divide them into HTTP/2 frames and to encrypt them (use of TLS is a de-facto requirement for HTTP/2).
The TLS part may need to be revisited if and when Netflix implements their "sendfile doing TLS", see slide 35 of this presentation (http://www.slideshare.net/facepalmtarbz2/new-sendfile-in-english)
Thanks to [BSD Now](http://www.bsdnow.tv/episodes/2015_01_12-get_your_engine_x_started) for making me aware of this development.
